macros/src/from_macro.rs

```diff
@@ -141,83 +156,294 @@ impl Register {
+
+        input.parse::<syn::Token![=]>()?;
+        let copy_of = input.parse()?;
+
+        let contents;
+        braced!(contents in input);
+
+        Ok(Self {
+            name,
+            description,
+            cfg_attributes,
+            kind: RegisterKind::Ref {
+                address_type: {
                     contents.parse::<syn::Token![const]>()?;
-                    contents.parse::<kw::RESET_VALUE>()?;
+                    contents.parse::<kw::BASE_ADDRESS>()?;
                     contents.parse::<syn::Token![:]>()?;
-                    let t = contents.parse::<syn::Type>()?;
+                    contents.parse()?
+                },
+                base_address: {
                     contents.parse::<syn::Token![=]>()?;
-                    let v = contents.parse::<syn::Expr>()?;
+                    let value = contents.parse::<syn::LitInt>()?.base10_parse()?;
                     contents.parse::<syn::Token![;]>()?;
-
-                    parse_reset_value(t, v)?
-                } else {
-                    None
-                }
+                    value
+                },
+                copy_of,
             },
-            description,
-            cfg_attributes,
-            fields: contents
-                .parse_terminated(<Field as syn::parse::Parse>::parse, syn::Token![,])?,
         })
     }
 }
 
+impl syn::parse::Parse for Register {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let attributes = syn::Attribute::parse_outer(input)?;
+
+        if input.peek(kw::register) {
+            Ok(Register::parse_register(input, attributes)?)
+        } else if input.peek(kw::block) {
+            Ok(Register::parse_block(input, attributes)?)
+        } else if input.peek(syn::Token![ref]) {
+            Ok(Register::parse_ref(input, attributes)?)
+        } else {
+            Err(syn::Error::new(
+                input.span(),
+                "Must be `register`, `command` or `buffer`",
+            ))
+        }
+    }
+}
+
+impl<'a> From<&'a Register> for device_driver_generation::Register {
+    fn from(r: &'a Register) -> Self {
+        Self {
+            name: r.name.to_string(),
+            description: r.description.clone(),
+            cfg_attributes: r.cfg_attributes.clone(),
+            kind: (&r.kind).into(),
+        }
+    }
+}
+
+enum RegisterKind {
+    Register {
+        rw_type: RWType,
+        address_type: syn::Ident,
+        address_value: u64,
+        size_bits_value: u64,
+        byte_order: Option<ByteOrder>,
+        reset_value: Option<ResetValue>,
+        fields: Punctuated<Field, syn::Token![,]>,
+    },
+    Block {
+        address_type: Option<syn::Ident>,
+        base_address: Option<u64>,
+        repeat: Option<RegisterRepeat>,
+        registers: Punctuated<Register, syn::Token![,]>,
+    },
+    Ref {
+        address_type: syn::Ident,
+        base_address: u64,
+        copy_of: syn::Ident,
+    },
+}
```

Comment:
The way you set this is might have me refactor it later. I don't know yet. But for now it's fine. I'd have made block (and maybe ref) its own thing. If this needs to be extended in the future, it's gonna be a bit difficult. But the refactor can be done then as well, so let's just leave this be.

Reply:
I guess it reflects the generation crate setup. So the same thing might come up at some point there too